Do not auto-set timezone, allow date

rly opened the pull request:
## Motivation

Fix #1859, fix #1843

1. Date strings (e.g., `"2024-04-10"` without the "T") can now be roundtripped, so if a user writes a `datetime.date` object, it will be written without the "T" and read as a `datetime.date` object. 
2. The local timezone is no longer automatically set on a non-timezone-aware datetime
3. timestamps_reference_time is allowed to be timezone-naive
4. Subject date_of_birth, file session_start_time, file file_create_date, and file timestamps_reference_time now allow date in addition to datetime


## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR clearly describes the problem and the solution?
- [x] Is your contribution compliant with our coding style? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [x] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

rly committed:
Do not auto-set timezone, allow date

rly committed:
Update docs

rly committed:
Add tests

rly committed:
Ruff and test fixes

rly commented:
@bendichter What do you think about this approach?

bendichter reviewed (dismissed):
Looks great!!! Do we still have tests that include timezones? (I'm on my phone)

rly commented:
Great, thanks. I removed the affected tests and added some tests to ensure that the following cases work:
1) datetime with time zone
2) datetime without time zone
3) date

I want to add a few more tests and remove timezone from a few tests, so this is not ready to merge yet.

bendichter commented:
ok. Looks great so far!

rly commented:
I think given a dataset spec or attribute spec with dtype isodatetime, the mapping of a string value to a datetime or date object should occur in HDMF on construction of the object, rather than as a custom constructor argument override on every field that we know has an isodatetime spec. Requiring the override to get the string as a datetime or date object creates a burden on any extension that wants to store datetimes and read datetime data as datetime objects instead of as strings.

So, `parse_datetime` should move to HDMF's `ObjectMapper`. I'll create an issue on HDMF. If it takes longer to implement that than the next pynwb release, we can just merge this change and remove it later once HDMF handles the mapping.

rly committed:
add docstring

rly committed:
Update CHANGELOG.md

rly committed:
Update CHANGELOG.md

rly committed:
Update CHANGELOG.md

rly committed:
Merge branch 'dev' into datetime

rly committed:
Merge branch 'datetime' of github.com:NeurodataWithoutBorders/pynwb into datetime

rly committed:
Update tests

rly committed:
Fix ruff
